<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Hacking IDclust • IDclust</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Hacking IDclust">
<meta property="og:description" content="IDclust">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IDclust</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.00</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Hackability.html">Hacking IDclust</a>
    </li>
    <li>
      <a href="../articles/scEpigenomics.html">Tutorial for scEpigenomics object (SingleCellExperiment)</a>
    </li>
    <li>
      <a href="../articles/scRNA.html">Tutorial for Seurat object (scRNA)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vallotlab/IDclust/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Hackability_files/header-attrs-2.14/header-attrs.js"></script><script src="Hackability_files/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1">
<link href="Hackability_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet">
<script src="Hackability_files/bootstrap-3.3.5/js/bootstrap.min.js"></script><script src="Hackability_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script><script src="Hackability_files/bootstrap-3.3.5/shim/respond.min.js"></script><style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="Hackability_files/navigation-1.1/tabsets.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Hacking IDclust</h1>
                        <h4 class="author">Pacome Prompsy</h4>
            
            <h4 class="date">5/17/2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vallotlab/IDclust/blob/master/vignettes/Hackability.Rmd"><code>vignettes/Hackability.Rmd</code></a></small>
      <div class="hidden name"><code>Hackability.Rmd</code></div>

    </div>

    
    
<p>In this article, we will show how to run IDclust with your own custom functions. The two ‘hackable’ part of IDclust are the processing and the differential functions. There are a few default functions present in the package, however based on your preferences you are able to customize both step quite easily.</p>
<p>As an example, we will analyse a scRNA dataset with <strong>‘monocle3’</strong> instead of the default Seurat for preprocessing as well as differential analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vallotlab/IDclust">IDclust</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">monocle3</span><span class="op">)</span></code></pre></div>
<div id="data" class="section level1">
<h1 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h1>
<p>Take example scRNA data from “Joint profiling of histone modifications and transcriptome in single cells from mouse brain,Chenxu Zhu, Yanxiao Zhang, Yang Eric Li, Jacinta Lucero, . Margarita Behrens, Bing Ren, Nature Methods, 2021 Paired-Tag”.<br>
We load the data and transform it into a cellDataset to analyze it with <strong>‘monocle3’</strong>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">47</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">Seu</span>, package <span class="op">=</span> <span class="st">"IDclust"</span><span class="op">)</span></code></pre></div>
</div>
<div id="defining-a-preprocessing-function" class="section level1">
<h1 class="hasAnchor">
<a href="#defining-a-preprocessing-function" class="anchor"></a>Defining a preprocessing function</h1>
<p>At each step of the ‘iterative_differential_clustering’ function, a cluster is re-processed. This means that the counts are re-normalized and that a new dimensionality reduction is calculated. Today we want to replace the default function by the processing used in ‘monocle3’ package.</p>
<p>We start by studying the ‘preprocess_Seurat’ function. It takes in entry 3 arguments:<br>
* ‘Seurat’ - A Seurat object * ‘n_dims’ - An integer specifying the number of first dimensions to keep in the dimensionality reduction step. * ‘dim_red’ - The name of the slot to save the dimensionality reduction at each step in the reducedDimNames(scExp).</p>
<p>It returns a Seurat object with the cell dimensionality reduction (embedding) stored in the ‘dim_red’ slot of the reducedDimNames of the object.</p>
<p>We therefore define a function that process the data using ‘monocle3’ package and matches the criteria describe above. We will have to convert the object to a ‘CellDataset’ object and back to a Seurat object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># From monocle3 tutorial -</span>
<span class="co"># https://cole-trapnell-lab.github.io/monocle3/docs/clustering/</span>

<span class="va">processing_monocle3</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Seu</span>, <span class="va">n_dims</span> <span class="op">=</span> <span class="fl">100</span>, <span class="va">dim_red</span> <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span> <span class="op">{</span>

    <span class="co"># Transform the Seurat object into CDS</span>
    <span class="va">gene_metadata</span> <span class="op">=</span> <span class="va">Seu</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">meta.features</span>
    <span class="va">gene_metadata</span><span class="op">$</span><span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">)</span>
    <span class="va">gene_metadata</span><span class="op">$</span><span class="va">gene_short_name</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">)</span>
    <span class="va">cds</span> <span class="op">=</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/new_cell_data_set.html">new_cell_data_set</a></span><span class="op">(</span>expression_data <span class="op">=</span> <span class="va">Seu</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>, cell_metadata <span class="op">=</span> <span class="va">Seu</span><span class="op">@</span><span class="va">meta.data</span>,
        gene_metadata <span class="op">=</span> <span class="va">gene_metadata</span><span class="op">)</span>

    <span class="co"># Process the cds</span>
    <span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html">preprocess_cds</a></span><span class="op">(</span><span class="va">cds</span>, num_dim <span class="op">=</span> <span class="va">n_dims</span><span class="op">)</span>
    <span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html">reduce_dimension</a></span><span class="op">(</span><span class="va">cds</span>, preprocess_method <span class="op">=</span> <span class="va">dim_red</span>, reduction_method <span class="op">=</span> <span class="va">dim_red</span><span class="op">)</span>

    <span class="co"># Transform back into a Seurat object and keep the metadata</span>
    <span class="va">Seu</span> <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span>, meta.data <span class="op">=</span> <span class="va">Seu</span><span class="op">@</span><span class="va">meta.data</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span>

    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">Seu</span><span class="op">)</span> <span class="op">=</span> <span class="va">Seu</span><span class="op">$</span><span class="va">seurat_clusters</span>

    <span class="co"># Add the reduced dimension</span>
    <span class="va">Seu</span><span class="op">@</span><span class="va">reductions</span><span class="op">[[</span><span class="va">dim_red</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateDimReducObject.html">CreateDimReducObject</a></span><span class="op">(</span>embeddings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDim</a></span><span class="op">(</span><span class="va">cds</span>,
        <span class="va">dim_red</span><span class="op">)</span>, key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">dim_red</span>, <span class="st">"_"</span><span class="op">)</span>, assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span>

    <span class="co"># Return the Seurat object</span>
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">Seu</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>We can now test the function, make sure that the object returned is a Seurat object and that the embedding is present in the object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Seu_monocle</span> <span class="op">=</span> <span class="fu">processing_monocle3</span><span class="op">(</span><span class="va">Seu</span>, n_dims <span class="op">=</span> <span class="fl">100</span>, dim_red <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">Seu_monocle</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Seurat"
## attr(,"package")
## [1] "SeuratObject"</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">Seu_monocle</span><span class="op">)</span></code></pre></div>
<p><img src="Hackability_files/figure-html/unnamed-chunk-4-1.png" width="672"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">Seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<p><img src="Hackability_files/figure-html/unnamed-chunk-4-2.png" width="672"></p>
</div>
<div id="defining-a-differential-function" class="section level1">
<h1 class="hasAnchor">
<a href="#defining-a-differential-function" class="anchor"></a>Defining a differential function</h1>
<p>We now define the differential function that will use ‘top_markers’ function from ‘monocle3’ package.</p>
<p>We start by studying the ‘differential_Seurat’ function. It takes in entry 3 mandatory arguments:</p>
<ul>
<li>‘Seurat’ - A Seurat object<br>
</li>
<li>‘by’ - A character specifying the column of clusters to compare<br>
</li>
<li>logFC.th - the log2 FoldChange threshold above which a gene is considered as significantly differential<br>
</li>
<li>qval.th - the adjusted p.value threshold below which a gene is considered as significantly differential</li>
</ul>
<p>The function returns a data.frame containing the marker genes for each cluster. It contains a ‘cluster’ column, a ‘gene’ column.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">differential_monocle3</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Seu</span>, <span class="va">by</span> <span class="op">=</span> <span class="st">"IDcluster"</span>, <span class="va">logFC.th</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span>, <span class="va">qval.th</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">{</span>

    <span class="co"># Transform the Seurat object into CDS</span>
    <span class="va">gene_metadata</span> <span class="op">=</span> <span class="va">Seu</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">meta.features</span>
    <span class="va">gene_metadata</span><span class="op">$</span><span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">)</span>
    <span class="va">gene_metadata</span><span class="op">$</span><span class="va">gene_short_name</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_metadata</span><span class="op">)</span>
    <span class="va">cds</span> <span class="op">=</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/new_cell_data_set.html">new_cell_data_set</a></span><span class="op">(</span>expression_data <span class="op">=</span> <span class="va">Seu</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>, cell_metadata <span class="op">=</span> <span class="va">Seu</span><span class="op">@</span><span class="va">meta.data</span>,
        gene_metadata <span class="op">=</span> <span class="va">gene_metadata</span><span class="op">)</span>

    <span class="co"># Differential analysis the cds</span>
    <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/top_markers.html">top_markers</a></span><span class="op">(</span><span class="va">cds</span>, group_cells_by <span class="op">=</span> <span class="va">by</span>, genes_to_test_per_group <span class="op">=</span> <span class="fl">200</span>,
        verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

    <span class="co"># Rename the column concerning gene and cluster</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">=</span> <span class="st">"gene"</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">=</span> <span class="st">"cluster"</span>

    <span class="co"># We just use the adjusted p.value threshold</span>
    <span class="va">res</span> <span class="op">=</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">marker_test_q_value</span> <span class="op">&lt;</span> <span class="va">qval.th</span><span class="op">)</span>, <span class="op">]</span>

    <span class="co"># Return the Seurat object</span>
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>We can now test the function, make sure that the object returned is a data.frame containing the marker genes of each cluster as well as the cluster column.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">=</span> <span class="fu">differential_monocle3</span><span class="op">(</span><span class="va">Seu</span>, by <span class="op">=</span> <span class="st">"seurat_clusters"</span>, qval.th <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
<pre><code>##      gene gene_short_name cluster marker_score mean_expression
## 1  Adarb2          Adarb2       1  0.190743353       4.7852141
## 2  Adarb2          Adarb2       4  0.300782439       4.0572739
## 4    Flt1            Flt1       3  0.099162684       1.0071785
## 5   Erbb4           Erbb4       0  0.006797023       0.2426385
## 6   Erbb4           Erbb4       1  0.613972449       8.3738044
## 8 Slco1a4         Slco1a4       3  0.098571404       0.6867022
##   fraction_expressing specificity  pseudo_R2 marker_test_p_value
## 1           0.4476987  0.42605291 0.09714333        2.654011e-24
## 2           0.8035714  0.37430703 0.32395038        1.393469e-68
## 4           0.1458333  0.67997269 0.06757809        1.030568e-15
## 5           0.1624266  0.04184673 0.10082267        4.556533e-36
## 6           0.8828452  0.69544747 0.60309324       2.648137e-165
## 8           0.1302083  0.75702838 0.07556654        2.025770e-17
##   marker_test_q_value
## 1        3.184814e-20
## 2        1.672163e-64
## 4        1.236682e-11
## 5        5.467840e-32
## 6       3.177764e-161
## 8        2.430924e-13</code></pre>
</div>
<div id="defining-a-differential-function-1" class="section level1">
<h1 class="hasAnchor">
<a href="#defining-a-differential-function-1" class="anchor"></a>Defining a differential function</h1>
<p>We can finally use our two function to run the ‘iterative_differential_clustering’ function. To do so, we pass our monocle3 customized functions to preprocessing_function and differential_function, and we make sure the dim_red and n_dims arguments are correctly set.</p>
<p>We first pre-process our object with our custom function :</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"Seu"</span><span class="op">)</span>
<span class="va">Seu_monocle</span> <span class="op">=</span> <span class="fu">processing_monocle3</span><span class="op">(</span><span class="va">Seu</span>, n_dims <span class="op">=</span> <span class="fl">100</span>, dim_red <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">47</span><span class="op">)</span>
<span class="va">output_dir</span> <span class="op">=</span> <span class="st">"~/Tests/IDC_monocle/"</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span><span class="op">(</span><span class="va">output_dir</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">output_dir</span><span class="op">)</span>

<span class="va">Seu_monocle</span> <span class="op">=</span> <span class="fu"><a href="../reference/iterative_differential_clustering.html">iterative_differential_clustering</a></span><span class="op">(</span><span class="va">Seu_monocle</span>, output_dir <span class="op">=</span> <span class="va">output_dir</span>,
    plotting <span class="op">=</span> <span class="cn">FALSE</span>, saving <span class="op">=</span> <span class="cn">TRUE</span>, n_dims <span class="op">=</span> <span class="fl">50</span>, dim_red <span class="op">=</span> <span class="st">"PCA"</span>, processing_function <span class="op">=</span> <span class="va">processing_monocle3</span>,
    differential_function <span class="op">=</span> <span class="va">differential_monocle3</span>, logFC.th <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span>, qval.th <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></code></pre></div>
<p>A ‘IDcluster’ column was added to the object, which we can retrieve and add to the original object. We can now plot the clusters found this way on the UMAP:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Seu</span><span class="op">$</span><span class="va">IDcluster</span> <span class="op">=</span> <span class="va">Seu_monocle</span><span class="op">$</span><span class="va">IDcluster</span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">Seu</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"IDcluster"</span><span class="op">)</span></code></pre></div>
<p><img src="Hackability_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>We can also plot the cluster network using the summary and tghe</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">IDC_summary</span> <span class="op">=</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html">qread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="st">"IDC_summary.qs"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_cluster_network.html">plot_cluster_network</a></span><span class="op">(</span><span class="va">Seu</span>, IDC_summary <span class="op">=</span> <span class="va">IDC_summary</span>, color_by <span class="op">=</span> <span class="st">"IDcluster"</span>, node_size_factor <span class="op">=</span> <span class="fl">4</span>,
    legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="Hackability_files/figure-html/unnamed-chunk-11-1.png" width="960"></p>
<pre><code>##    IDcluster IDcluster_color
## 1      A1:B1           plum1
## 2   A1:B2:C1          ivory4
## 3   A1:B2:C2     forestgreen
## 4   A1:B3:C1      violetred1
## 5   A1:B3:C2      palegreen3
## 6      A1:B4    navajowhite3
## 7      A1:B5      lightblue2
## 8      A1:B6  palevioletred1
## 9   A2:B1:C1 darkolivegreen2
## 10  A2:B1:C2      olivedrab2
## 11     A2:B2       seashell1
## 12        A3         purple1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Pacome Prompsy, Celine Vallot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
